core/shared/src/main/scala/zio/prelude/fx/ImperativeDsl.scala

```diff
@@ -0,0 +1,181 @@
+              }
+            case Right(a) =>
+              stack match {
+                case ImperativeDsl.Sequence(_, onSuccess, _) :: stack => loop(onSuccess(a), stack)
+                case Nil                                              => Right(a.asInstanceOf[A])
+              }
+          }
+        case free @ ImperativeDsl.Sequence(fa, _, _) =>
+          loop(fa, (free :: stack).asInstanceOf[List[ImperativeDsl.Sequence[Dsl, Any, Any, Any, Any]]])
+      }
+    loop(self, Nil)
+  }
+}
+
+object ImperativeDsl {
+  def eval[Op[+_, +_], E, A](fa: Op[E, A]): ImperativeDsl[Op, E, A] = Eval(fa)
+  def fail[Op[+_, +_], E](e: E): ImperativeDsl[Op, E, Nothing]      = Opail(e)
+  def succeed[Op[+_, +_], A](a: A): ImperativeDsl[Op, Nothing, A]   = Succeed(a)
+
+  final case class Succeed[Op[+_, +_], A](a: A)         extends ImperativeDsl[Op, Nothing, A]
+  final case class Opail[Op[+_, +_], E](a: E)           extends ImperativeDsl[Op, E, Nothing]
+  final case class Eval[Op[+_, +_], E, A](fa: Op[E, A]) extends ImperativeDsl[Op, E, A]
+  final case class Sequence[Op[+_, +_], E1, E2, A1, A2] private[ImperativeDsl] (
+    fa: ImperativeDsl[Op, E1, A1],
+    onSuccess: A1 => ImperativeDsl[Op, E2, A2],
+    onFailure: E1 => ImperativeDsl[Op, E2, A2]
+  ) extends ImperativeDsl[Op, E2, A2] {
+    type InSuccess = A1
+    type InFailure = E1
+  }
+
+  /// Interpreter provides the ability to interpret a DSL into an executable program
+  trait Interpreter[Dsl[+_, +_], Executable[+_, +_]] { self =>
+    def interpret[E, A](dsl: Dsl[E, A]): Executable[E, A]
+
+    def combine[Dsl2[+_, +_]](
+      that: Interpreter[Dsl2, Executable]
+    ): Interpreter[({ type lambda[+E, +A] = CompositeDsl[Dsl, Dsl2, E, A] })#lambda, Executable] =
+      new Interpreter[({ type lambda[+E, +A] = CompositeDsl[Dsl, Dsl2, E, A] })#lambda, Executable] {
+        override def interpret[E, A](dsl: CompositeDsl[Dsl, Dsl2, E, A]): Executable[E, A] = dsl.eitherDsl match {
+          case Left(dsl)  => self.interpret(dsl)
+          case Right(dsl) => that.interpret(dsl)
+        }
+      }
+  }
+
+  trait UnsafeInterpreter[Dsl[+_, +_]] {
+    def interpret[E, A](fa: Dsl[E, A]): Either[E, A]
+  }
+
+  trait ToExecutable[Executable[+_, +_]] {
+    def succeed[A](a: A): Executable[Nothing, A]
+    def fail[E](e: E): Executable[E, Nothing]
+    def eval[E, A](fa: Executable[E, A]): Executable[E, A]
+    def sequence[E1, E2, A1, A2](
+      fa: Executable[E1, A1],
+      onSuccess: A1 => Executable[E2, A2],
+      onFailure: E1 => Executable[E2, A2]
+    ): Executable[E2, A2]
+  }
+
+  final case class CompositeDsl[+Dsl1[+_, +_], +Dsl2[+_, +_], +E, +A](eitherDsl: Either[Dsl1[E, A], Dsl2[E, A]])
+      extends AnyVal { self =>
+    type InSuccess <: A
+    type InFailure <: E
+  }
+
+  // TODO: Consider what can be done to make the type lambda here simpler
+  implicit def ZPureToExecutable[W]
```

Comment:
No point to the `W` parameter here unless you want to add other cases to `Imperative`. There is no way to interact with it.